[ErrorHandler] Add helper script to patch type declarations

wouterj opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | -
| License       | MIT
| Doc PR        | https://github.com/symfony/symfony-docs/pull/15802

While trying out the patch functionality in my bundle (and simultaneously writing docs for it), I found that I would pretty much recommend to copy-paste the `.github/patch-types.php` script. So I made it a bit more generic and added it as an official script to the ErrorHandler component.

Usage:

```
# important!
$ composer dump-autoload -o

# default usage
$ ./vendor/bin/patch-type-declarations

# if you're an OSS maintainer, you probably want:
$ SYMFONY_PATCH_TYPE_DECLARATIONS="force=1&php=7.4" ./vendor/bin/patch-type-declarations

# regex if you need to exclude a certain directory or class (defaults to "vendor/" by default)
$ SYMFONY_PATCH_TYPE_EXCLUDE="{/vendor/|/path/to/excluded/}" ./vendor/bin/patch-type-declarations
```

For reference: I did not remove the `.github/patch-type.php` script. I don't think it's worth replacing all the exclusion rules in that script with a single regex.

@nicolas-grekas please let me know if you think this makes sense (or if I'm interpreting the new functionality completely wrong).

wouterj committed:
[ErrorHandler] Add helper script to patch type declarations

nicolas-grekas commented on src/Symfony/Component/ErrorHandler/Resources/bin/patch-type-declarations:
```diff
@@ -0,0 +1,13 @@
+ * file that was distributed with this source code.
+ */
+
+require __DIR__.DIRECTORY_SEPARATOR.'patch-type-declarations.php';
```
let's use `/` instead of `DIRECTORY_SEPARATOR`

but I'm not sure we need to split the .php file from this one. Can we drop the .php one and merge it's content here? (and set this file in 755 mode)

wouterj replied:
Yes, that works for me. I just replicated the `simple-phpunit` set-up :) (including directory separator)

nicolas-grekas commented on src/Symfony/Component/ErrorHandler/Resources/bin/patch-type-declarations.php:
```diff
@@ -0,0 +1,43 @@
+    exit(1);
+}
+
+if (is_file($phpunitAutoload = pathinfo($autoload, \PATHINFO_DIRNAME).'/bin/.phpunit/phpunit/vendor/autoload.php')) {
```
why not using `dirname()`?

nicolas-grekas commented on src/Symfony/Component/ErrorHandler/Resources/bin/patch-type-declarations.php:
```diff
@@ -0,0 +1,43 @@
+
+if (false === getenv('SYMFONY_PATCH_TYPE_DECLARATIONS')) {
+    putenv('SYMFONY_PATCH_TYPE_DECLARATIONS=force=2');
+    echo 'Patching type declarations in all methods';
```
missing `\PHP_EOL` to all echo lines

nicolas-grekas commented on src/Symfony/Component/ErrorHandler/Resources/bin/patch-type-declarations.php:
```diff
@@ -0,0 +1,43 @@
+
+Symfony\Component\ErrorHandler\DebugClassLoader::enable();
+
+$exclude = getenv('SYMFONY_PATCH_TYPE_EXCLUDE') ?: '/'.preg_quote(\DIRECTORY_SEPARATOR.'vendor'.\DIRECTORY_SEPARATOR).'/';
```
It'd be great to document `SYMFONY_PATCH_TYPE_EXCLUDE` right into this file, maybe the rest a bit too.

nicolas-grekas replied:
note that vendor files are already ignored by `DebugClassLoader::checkClasses()`, see
https://github.com/symfony/symfony/blob/64cbfd2569840777a54b849bb52a42aae084417c/src/Symfony/Component/ErrorHandler/DebugClassLoader.php#L532

nicolas-grekas reviewed (commented):
Thanks a lot for this PR, I had something similar on my list!

wouterj commented:
PR updated, including the proposed changes. Also added some help information, and added a basic deprecation handling (I don't want to depend on PhpUnitBridge - but it's still good to show type deprecations imho).

nicolas-grekas commented on src/Symfony/Component/ErrorHandler/Resources/bin/patch-type-declarations:
```diff
@@ -0,0 +1,86 @@
+
+if (\in_array('-h', $argv) || \in_array('--help', $argv)) {
+    echo implode(PHP_EOL, [
+        ' Patches type declarations to fix incompatibility with Symfony 6.',
```
Note that the script isn't bound to Symfony 6. It works for all deps in vendor/ (that might be important for the doc also: this is part or a generic deprecation framework for return types)

nicolas-grekas commented on src/Symfony/Component/ErrorHandler/Resources/bin/patch-type-declarations:
```diff
@@ -0,0 +1,86 @@
+    echo implode(PHP_EOL, [
+        ' Patches type declarations to fix incompatibility with Symfony 6.',
+        '',
+        ' Available options:',
```
Available configuration via environment variables:

derrabus reviewed (approved):
Thank you!

As a follow-up, I think it would be great if we could eat our own dogfood here and leverage that script of yours in our CI.

derrabus commented:
Thank you Wouter.
